src/main/kotlin/com/github/burrunan/s3cache/internal/AwsS3BuildCacheServiceFactory.kt

```diff
@@ -69,51 +70,46 @@ class AwsS3BuildCacheServiceFactory : BuildCacheServiceFactory<AwsS3BuildCache>
         check(!config.bucket.isNullOrEmpty()) { "S3 build cache has no bucket configured" }
     }
 
-    private fun createS3Client(config: AwsS3BuildCache) = AmazonS3ClientBuilder.standard().run {
+    private fun createS3Client(config: AwsS3BuildCache) = S3AsyncClient.builder().run {
         addHttpHeaders(config)
         addCredentials(config)
-        if (config.endpoint.isNullOrEmpty()) {
-            withRegion(config.region)
-        } else {
-            withEndpointConfiguration(
-                AwsClientBuilder.EndpointConfiguration(
-                    config.endpoint,
-                    config.region
-                )
-            )
+        region(Region.of(config.region))
+        config.endpoint?.let {
+            val endpoint = if (it.startsWith("http")) it else "https://${it}"
+            endpointOverride(URI.create(endpoint))
         }
-        enablePathStyleAccess()
         build()
     }
 
-    private fun AmazonS3ClientBuilder.addHttpHeaders(
+    private fun S3AsyncClientBuilder.addHttpHeaders(
         config: AwsS3BuildCache
     ) {
-        clientConfiguration = ClientConfiguration().apply {
-            for ((key, value) in config.headers ?: mapOf()) {
-                if (key != null && value != null) {
-                    addHeader(key, value)
+        config.headers?.let { headers ->
+            overrideConfiguration {
+                for ((key, value) in config.headers ?: mapOf()) {
+                    if (key != null && value != null) {
+                        it.putHeader(key, value)
+                    }
                 }
             }
         }
     }
 
-    private fun AmazonS3ClientBuilder.addCredentials(config: AwsS3BuildCache) {
+    private fun S3AsyncClientBuilder.addCredentials(config: AwsS3BuildCache) {
         val credentials = when {
-            config.awsAccessKeyId.isNullOrBlank() && config.awsSecretKey.isNullOrBlank() -> when {
+            config.awsAccessKeyId.isNullOrBlank() || config.awsSecretKey.isNullOrBlank() -> when {
```

Comment:
Noticed I've changed the conditions here. If either key is missing, we can't provide a credential so `||` seems good here. And below, I couldn't see any reason to check whether access key is empty when creating basic creds, it seems to make more sense to check session token.